### DISCUSSION TITLE
Proper setup with ZSH?

### DISCUSSION THREAD

**Discussion by 9mm** [neovide/neovide#2541] [Q&A]
I found this comment: https://github.com/neovide/neovide/issues/950#issuecomment-1671100576

I've been trying to fix this for like 3 hours now and I am losing my mind. I realize this really has nothing to do with Neovide... moreso ZSH. Thats why I put it in discussions as I dont really expect much.

It started when I realized all the language providers were picking up the wrong version (system ruby) inside neovide, but it worked in neovim.

I then found that the issue is because of my `$PATH`.

It's hard to explain what the issue is as I've tried so many things, which have lots of... side effects to the explanation, but the simple answer is... I'm ultimately trying to figure out why `/usr/bin` is appearing at the front of my path, which is causing Neovide to read all the wrong binaries. Neovim actually gets the correct binaries.

I'm also trying to do this in a way thats the proper way, instead of just loading the path tons of duplicate times just to get it working.

There is a secondary issue where `.zprofile` is loaded twice which is super confusing.

**.zprofile**

```
export PATH="/ZZZZZZZZZZZZZZZPROFILE:$PATH"
eval "$(/opt/homebrew/bin/brew shellenv)"
```

The eval line returns this:

```
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```

**.zshenv**

```
export PATH="/ZZZZZZZZZZZZZZZENV:$PATH"
```

**.zshrc**

I actually don't have anything in here except a bunch of exports (im NOT modifying the path except for language specific stuff). i was going to move them per @fredizzimo advice in the link above until i realized somehow neovide IS loading them even though theyre inside .zshrc. Heres a few lines just so you can see them appear below:

```
# Postgres
export PATH="/Applications/Postgres.app/Contents/Versions/16/bin:$PATH"

# Ruby
eval "$(rbenv init - zsh)"

# Go
export GOPATH="$HOME/.go"
export PATH="$GOPATH/bin:$PATH"
```

**EXPECTED OUTPUT for $PATH (Neovim running in terminal)**

```
/Users/zesty/.local/share/nvim/mason/bin:
/Users/zesty/Library/pnpm:
/Users/zesty/.go/bin:
/Users/zesty/.rbenv/shims:
/Applications/Postgres.app/Contents/Versions/16/bin:
/Users/zesty/.local/bin:
/Users/zesty/.zsh/bin:
/opt/homebrew/bin:
/opt/homebrew/sbin:
/ZZZZZZZZZZZZZZZPROFILE:
/usr/local/bin:
/System/Cryptexes/App/usr/bin:
/usr/bin:
/bin:
/usr/sbin:
/sbin:
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:
/Applications/Little Snitch.app/Contents/Components:
/ZZZZZZZZZZZZZZZENVAFTER:
/ZZZZZZZZZZZZZZZENV:
/Applications/iTerm.app/Contents/Resources/utilities:
/Users/zesty/go/bin
```

**ACTUAL OUTPUT FOR $PATH** (Neovide `echo $PATH` command)

```
/ZZZZZZZZZZZZZZZENV:
/Users/zesty/.local/share/nvim/mason/bin:
/opt/homebrew/bin:
/opt/homebrew/sbin:
/ZZZZZZZZZZZZZZZPROFILE:
/usr/local/bin:
/System/Cryptexes/App/usr/bin:
/usr/bin:
/bin:
/usr/sbin:
/sbin:
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:
/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:
/Applications/Little Snitch.app/Contents/Components:
/ZZZZZZZZZZZZZZZENV:
/Users/zesty/Library/pnpm:
/Users/zesty/.go/bin:
/Users/zesty/.rbenv/shims:
/Applications/Postgres.app/Contents/Versions/16/bin:
/Users/zesty/.local/bin:
/Users/zesty/.zsh/bin:
/opt/homebrew/bin:
/opt/homebrew/sbin:
/ZZZZZZZZZZZZZZZPROFILE:
/Applications/iTerm.app/Contents/Resources/utilities:
/Users/zesty/go/bin
```

**Comment by 9mm** [comment]
Ugh, I realize `echo $PATH` is different than `!echo $PATH` now...

This is too confusing to solve so I decided to just prepend my paths in `.zshenv` and be done with it. Then i realize this DOES show the correct path:

`!echo $PATH`

however this does NOT (/usr/bin is still at the front of the list).. fuuuuuuu

`echo $PATH`

**Reply by fredizzimo** [reply]
I think this might actually be a bug introduced by https://github.com/neovide/neovide/pull/2405.

It seems very plausible, since it starts two shells. But I have not done a full analysis yet.

**Reply by fredizzimo** [reply]
Yes, alacritty only executes the login shell once https://github.com/alacritty/alacritty/blob/7077a5f4cb5d23ce91a0509f7fae423458698170/alacritty_terminal/src/tty/unix.rs#L180 (`-c` instead of `-lc`)

**Reply by fredizzimo** [reply]
I even commented about it during the review https://github.com/neovide/neovide/pull/2405#discussion_r1518355410

**Reply by fredizzimo** [reply]
I also think it might be missing a check that prevents login from being run, if the user is akready logged in, like when running from the terminal

**Reply by 9mm** [reply]
Haha... Im somewhat happy this is a potential bug, cause I literally was thinking I am crazy? I dont deal with nuances of all the shells that often but the amount of weirdness was extremely hard to explain or reason about.

**Reply by fredizzimo** [reply]
You can try changing the code to more like it was before. This login behaviour is only to fix the case where you are not running from a terminal. And, I'm not sure how things inherit when you are forking.

**Reply by falcucci** [reply]
Running the shell removing the `-l` caused issues about not loading the correct `$PATH` as well, that's the reason why I have forced as you can see in my comment https://github.com/neovide/neovide/pull/2405#discussion_r1518400021 @fredizzimo

**Reply by 9mm** [reply]
@falcucci do you have a suggestion how to fix this current issue then?

**Reply by falcucci** [reply]
If this is still an issue, I would then vote to rollback also this particular change, running the `$shell` without the `/usr/bin/login` as before.

**Reply by falcucci** [reply]
@9mm also to troubleshoot try to load the path and validate your environment as neovide executes the embedded neovim instance for the bridge communication, as:

```bash
"/usr/bin/login" "-flpq" "<user>" "<shell>" "-lc" "echo $PATH"
```

**Reply by 9mm** [reply]
If I run this from within a zsh terminal window, I see this output, ... which seems correct:

<img width="970" alt="image" src="https://github.com/neovide/neovide/assets/1386207/9708477d-e045-45da-b9f3-2cdfb4a94dc3">

Here is the output however of `:echo $PATH`:

<img width="865" alt="image" src="https://github.com/neovide/neovide/assets/1386207/588cc613-8459-4266-8ac2-6af061803bcf">


and `:!echo $PATH` (with a ! sign)... not sure which is correct

<img width="875" alt="image" src="https://github.com/neovide/neovide/assets/1386207/ca0341d8-1e33-4e33-957e-884f32f42257">

Sorry about the screenshots it is easier than getting pbcopy to work in there

**Reply by 9mm** [reply]
Note that...

- theyre all different order
- zshenv + zprofile both load twice in 2nd one
- zprofile loads twice in the first one
- zshrc is within my $PATH, I have no idea how that is happening according the flowchart that @fredizzimo put which according to that, neovide shouldnt be loading. i could be doing something dumb. this is in the file:

<img width="578" alt="image" src="https://github.com/neovide/neovide/assets/1386207/c181415b-331b-4fca-92f9-fec6681e7e61">

Maybe it has to do with something fancy brew is doing, I'm not sure. On my M1 mac, this is what people said to add to `.zprofile`:

```
eval "$(/opt/homebrew/bin/brew shellenv)"
```

I searched around on SO for some time to see if this was correct or just the lazy answer... it would appear lots of people say the same thing, but that doesnt necessarily mean its the right way.

That script evaluates to:

```
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```

I honestly have *NO* idea what is "expected" or not, all i know is i stare at this like truly wtf

**Reply by falcucci** [reply]
are you launching neovide from the terminal or the Dock/Finder?

**Reply by 9mm** [reply]
Great question, I almost forgot about that. I use a launcher script I wrote (which i run from terminal) so I can control the working directory of neovim without having the "file picker" popup open by default, and some other things.

With that said, I tried the above commands directly without script by simply launching like this from command line:

```
NEOVIDE_FORK=1 ~/.cargo/bin/neovide
```

and I get the same results, so luckily my launcher script hopefully didnt add any side effects

I should also mention that I do launch Neovide from the dock rarely, but I haven't even tested that at all yet

```bash
export NEOVIDE_FRAME=buttonless
export NEOVIDE_TITLE_HIDDEN=1
export NEOVIDE_VSYNC=0
export NEOVIDE_FORK=1

neovide_bin="/Users/zesty/.cargo/bin/neovide" # "/opt/homebrew/bin/neovide"

current_dir=$(pwd)

if [ -z "$1" ]; then
  if [ "$current_dir" = "$HOME" ]; then
    $neovide_bin -- -c ":cd ~/Desktop"
  else
    $neovide_bin -- -c ":cd $current_dir"
  fi
else
  if [ -d "$1" ]; then
    $neovide_bin -- -c ":cd $1"
  else
    $neovide_bin "$@"
  fi
fi
```

**Reply by 9mm** [reply]
Oh wow, if I launch it from the dock i get even different results:

Oooooh I see.. so maybe when I launch it from terminal its loading everything again, and the dock app is not inheriting from my terminal?

`:echo $PATH`

<img width="910" alt="image" src="https://github.com/neovide/neovide/assets/1386207/63ca4b91-fdd4-4826-a85d-11075ed9eb8b">

`:!echo $PATH`

<img width="887" alt="image" src="https://github.com/neovide/neovide/assets/1386207/ce2a7b2d-ec0a-472e-87eb-e95e77557b90">

**Reply by falcucci** [reply]
so maybe that's a thing for the `tty` condition to fix.

The problem is that on macOS @fredizzimo has pointed out a bug on rust side to [identify](https://doc.rust-lang.org/nightly/std/io/trait.IsTerminal.html) it, as a false negative, which we are not able to avoid the `-l` for such cases when needed *properly*. Maybe we could use a hack here.

In theory this behavior shouldn't happen due to the neovide concept of being (or trying to be) always identified as [non-tty](https://github.com/neovide/neovide/blob/fcf5e87bf8a75727ee75da961ce98b2f51fd0cb2/src/main.rs#L239-L248) as default.

**Reply by fredizzimo** [reply]
@falcucci, I don't think I ever identified that as a bug, and I can't find any reports online that indicates it is. So, I think we should assume that `is_terminal()` works.

If the identification really is wrong, then the `--fork` option would not work at all, since it would break out here before it has a chance to fork
https://github.com/neovide/neovide/blob/35ea46a3fef7f9938830dd8e294b696dfe24af56/src/main.rs#L242-L243

I believe that @9mm, identified the core issue here, when you launch from the terminal, the user is already logged in, so Neovide should not perform a login again. Forking/non-forking makes no difference here, since the environment is inherited in the forked process. 

That's the reason why Alacritty uses `exec -a`, and why we have to do the same it cleans the environment, so it's safe to start a login shell. 
https://ss64.com/bash/exec.html

The exec command is not available in all shells though, so that's why they hardcode zsh (which should be available on all macOS systems) for launching `exec`, which internally launches the user shell.

Note that this method of launching does not need to check if launched from a terminal or not, since the environment is always cleaned.

Also note that this cryptic line `let exec = format!("exec -a -{} {}", shell_name, shell);` causes the launched shell to become a login shell due to the `-`.  Actually, it passes `-shell_name` as the first argument to the shell.  For example `zsh` deals with that here https://github.com/zsh-users/zsh/blob/f6e005a9ef55b1fbc5291729faa2ab46bf3d9e8f/Src/init.c#L260-L261, and makes it a login shell. 

And finally note, that the `-l` parameter is passed to `login` so that it does not run a login shell automatically, as that could again lead to the same duplication https://ss64.com/mac/login.html

So, as a conclusion, my suggestion is to change the code to exactly how it's done in Alacritty.

**Reply by falcucci** [reply]
@fredizzimo at our discussions at discord, the `is_terminal()` implementation wasn't behaving as expected according to you. That was the main reason to make the rollback about the startup feature. And also some regressions, but the idea itself felt there and wasn't worth the effort for any fixes.

About Alacritty: I've done this as the same as they do in the past and it didn't load the `$PATH` correctly for non-tty use cases but maybe I was doing something wrong tho. @9mm feel free to replicate that again if you have some time, launching from terminal will probably work. Must be tested launching from Dock/Finder well afterwards.

just update the code at https://github.com/neovide/neovide/blob/5efb713ded6435c7384ab5ba395903a3b491d2fa/src/bridge/command.rs?plain=1#L74

**Reply by fredizzimo** [reply]
> @fredizzimo at our discussions at discord, the is_terminal() implementation wasn't behaving as expected according to you. That was the main reason to make the rollback about the startup feature. And also some regressions, but the idea itself felt there and wasn't worth the effort for any fixes.

Yes, but I suspect that there was something in the way the testing was done that was the problem. But it is possible that it is some machine specific issue, since for @9mm, it does seem to report correctly, or at least the forking option works.

@falcucci, would it be possible for you to test Alacritty, launching from the dock? That's probably the easiest way to tell if their approach works or not.

**Reply by 9mm** [reply]
```
let cmd_args = vec!["-flpq", &user, &shell, "-lc", &exec]; 
```

If im testing this on neovide, what am i updating this to? That's what it is already set to

**Reply by fredizzimo** [reply]
The same as alacritty, both the exec command and the actual command https://github.com/alacritty/alacritty/blob/7077a5f4cb5d23ce91a0509f7fae423458698170/alacritty_terminal/src/tty/unix.rs#L166-L180

```rust
    let shell_name = shell.rsplit('/').next().unwrap();

    // On macOS, use the `login` command so the shell will appear as a tty session.
    let mut login_command = Command::new("/usr/bin/login");

    // Exec the shell with argv[0] prepended by '-' so it becomes a login shell.
    // `login` normally does this itself, but `-l` disables this.
    let exec = format!("exec -a -{} {}", shell_name, shell);

    // -f: Bypasses authentication for the already-logged-in user.
    // -l: Skips changing directory to $HOME and prepending '-' to argv[0].
    // -p: Preserves the environment.
    //
    // XXX: we use zsh here over sh due to `exec -a`.
    login_command.args(["-flp", user, "/bin/zsh", "-c", &exec]);
```

**Reply by 9mm** [reply]
```rust
    let user = env::var("USER").unwrap_or_explained_panic("USER environment variable not found");
    let shell = env::var("SHELL").unwrap_or_else(|_| "/bin/zsh".to_string());

    let args = match shlex::try_join(args.iter().map(|s| (*s) as &str)) {
        Ok(args) => args,
        Err(_) => panic!("Failed to join arguments"),
    };

    let exec = format!("exec -a -{} {}", command, args);

    let cmd_path = "/usr/bin/login";

    let cmd_args = vec!["-flp", &user, &shell, "-c", &exec];
```

Can you help me see what im missing? I didnt copy alacritty exact as I see it seems they're passing in different arguments and stuff. I just changed mostly the -flp command and the format! macro contents

```
ERROR: Unexpected output from neovim binary:
	/opt/homebrew/bin/nvim -v
stdout:
stderr: zsh:1: exec requires a command to execute
```

**Reply by fredizzimo** [reply]
`let exec = format!("exec -a -{} {}", command, args);` this line needs the shell name as the first argument (just the name) for example `zsh` while I think the second argument should be something like `/usr/bin/zsh -c nvim -v`

**Reply by 9mm** [reply]
like this? now it compiles but when it launches, nothing actually opens. Sorry to be pain

i removed the format and stuff just for clear visibility into debugging

```
    let exec = "exec -a -zsh /bin/zsh -c /opt/homebrew/bin/nvim -v";

    let cmd_args = vec!["-flp", &user, &shell, "-c", &exec];
```

**Reply by fredizzimo** [reply]
I think that is correct but nvim will be called a second time with different arguments.

**Reply by nkarl** [reply]
You can do following:

```lua
if vim.g.neovide then
  local home = vim.env.HOME
  vim.env.PATH =
    home .. '/path/foo:' ..
    home .. '/path/bar:' .. vim.env.PATH
end
```

This is only a quick fix though. You will have to append extra paths when you have more.

**Comment by carlaarchuleta707-lgtm** [comment]
[    let exec = "exec -a -zsh /bin/zsh -c /opt/homebrew/bin/nvim -v";

    let cmd_args = vec!["-flp", &user, &shell, "-c", &exec];](url)
